<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [75]:
import pandas as pd
import psycopg2

In [76]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [77]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [67]:
# SQL-запрос для подсчета общего количества вакансий в таблице vacancies
# count(*) агрегатная функция PostgreSQL, которая подсчитывает количество строк
# Результат запроса- одно число с общим количеством записей в таблице
query_3_1 = f'''
select count(*) 
from vacancies  
'''

In [68]:
display(f'Ваканский в базе: {int(pd.read_sql_query(query_3_1, connection).iloc[0,0])}')

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_12092\3605000545.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(f'Ваканский в базе: {int(pd.read_sql_query(query_3_1, connection).iloc[0,0])}')


'Ваканский в базе: 49197'

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [69]:
#Подсчет кол-во работодателей
query_3_2 = f'''
select count(*)
from employers'''

In [70]:
display(f'Количество работодателей: {pd.read_sql_query(query_3_2, connection).iloc[0,0]}')

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_12092\2541219557.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(f'Количество работодателей:{pd.read_sql_query(query_3_2, connection).iloc[0,0]}')


'Количество работодателей:23501'

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [71]:
# SQL-функция для подсчета регионов в таблице areas
query_3_3 = f''' 
select count(*)
from areas '''

In [72]:
display(f'Количество регионов: {pd.read_sql_query(query_3_3, connection).iloc[0,0]}')

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_12092\499098353.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(f'Количество регионов:{pd.read_sql_query(query_3_3, connection).iloc[0,0]}')


'Количество регионов:1362'

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [73]:
# sql-запрос на подсчет сфер деятельности, в таблице industries
query_3_4 =f''' 
select count(*)
from industries'''

In [74]:
display(f'Количество сфер деятельности в базе: {pd.read_sql_query(query_3_4, connection).iloc[0,0]}')

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_12092\907306582.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(f'Количество сфер деятельности в базе:{pd.read_sql_query(query_3_4, connection).iloc[0,0]}')


'Количество сфер деятельности в базе:294'

***

### Выводы по предварительному анализу данных: 
- Ваканский в базе: 49197
- Количество работодателей: 23501
- Количество регионов: 1362
- Количество сфер деятельности в базе: 294

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [33]:
query_4_1 = ''' 
select count(v.name) as cnt, -- Подсчет количества вакансий,
       a.name as area               --Название региона
from vacancies v             --таблица vacancies
join areas a on a.id = v.area_id --inner join с таблицей регионов по ключу area_id
group by a.name              --Группировка по названию регионов
order by cnt DESC            --Сортировка по убыванию количества ваканский
limit 5                      --Ограничение вывода до 5 первых результатов
'''

In [36]:
display(f'Количество ваканский в каждом регионе:')
q_4_1 = pd.read_sql_query(query_4_1, connection)
q_4_1

'Количество ваканский в каждом регионе:'

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_2240\2263043766.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  q_4_1 = pd.read_sql_query(query_4_1, connection)


cnt             area
0  5333           Москва
1  2851  Санкт-Петербург
2  2112            Минск
3  2006      Новосибирск
4  1892           Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [38]:
# Запрос для подсчета количества вакансий, у которых указана зарплата (хотя бы одно из полей salary_from или salary_to заполнено)
query_4_2 = ''' 
select count(*)
from vacancies v 
where salary_from is not null or salary_to is not null
'''

In [40]:
display(f'Количество ваканский у которых указана ЗП(хотя бы одно из полей):{pd.read_sql_query(query_4_2, connection).iloc[0,0]}')

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_2240\2668680222.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(f'Количество ваканский у которых указана ЗП(хотя бы одно из полей):{pd.read_sql_query(query_4_2, connection).iloc[0,0]}')


'Количество ваканский у которых указана ЗП(хотя бы одно из полей):24073'

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [41]:
# Среднее значение нижней(salary_from) и верхней(salary_to) границы зарплатной вилки. Целое значение
query_4_3 ='''
select round(avg(v.salary_from)), round(avg(v.salary_to))
from vacancies v
'''

In [46]:
result = pd.read_sql_query(query_4_3,connection)
avg_from =  result.iloc[0,0]
avg_to = result.iloc[0,1]
display(f'Среднее значение нижней зарплатной вилки:{avg_from:.2f}')
display(f'Среднее значение нижней зарплатной вилки:{avg_to:.2f}')


C:\Users\arinathebest\AppData\Local\Temp\ipykernel_2240\1523601015.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  result = pd.read_sql_query(query_4_3,connection)


'Среднее значение нижней зарплатной вилки:71065.00'

'Среднее значение нижней зарплатной вилки:110537.00'

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
# Запрос для анализа распределения вакансий по комбинациям графика работы и типа трудоустройства
query_4_4 ='''
select schedule, --Тип рабочего графика 
       employment, --Тип трудоустройства
       count(*) as cnt --Кол-во ваканский
from vacancies
group by schedule, employment --Группировка по типу рабочего графика и типу трудоустройства
order by cnt desc --Отсортировать по количеству ваканский
'''

In [ ]:
df = pd.read_sql_query(query_4_4,connection)
df

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_12020\1742807029.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1,connection)


schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [ ]:
# Запрос выводящий значения Требуемый опыт работы в порядке возрастания вакансий 
query_4_5 = '''
select count(*) as cnt, --кол-во вакансий
       experience --Требуемый опыт работы
from vacancies
where experience is not null --фильтрация по вакансиям в которых указан опыт работы
group by experience --Группировка по опыту работы
order by cnt  --Сортировка по убыванию количеству вакансий
'''

In [119]:
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_12020\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


cnt          experience
0   1337         Более 6 лет
1   7197           Нет опыта
2  14511       От 3 до 6 лет
3  26152  От 1 года до 3 лет

***

### Выводы по детальному анализу вакансий:
-Большинсво вакансий находятся в Москве

-Практически у половины вакансий отсутвует данные о ЗП 

-ЗП варьируется в промежутке от 71тыс до 110тыс

-Абсолютное большинсво хотят работать полный день, полную занятость. На втором месте удаленная работа-Полная занятость

-Первое место по количеству требуемого опыта работы - от 1 года до 3х лет 

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [54]:
query_5_1 = '''
select e.name as emp_name, --Название работодатяеля
       count(*) as cnt --кол-во вакансий
from employers e 
join vacancies v on v.employer_id = e.id --inner join по ID-работодателя из таблицы vacancies = ID-работодателя из таблицы employers
group by emp_name --группировка по назанию работодателя
order by cnt DESC --Сортировка кол-во вакансий по убыванию
limit 5 -- Вывод 5 первых строк
'''

In [55]:
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_2240\1792814825.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


emp_name   cnt
0         Яндекс  1933
1     Ростелеком   491
2       Тинькофф   444
3           СБЕР   428
4  Газпром нефть   331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
# Запрос для поиска региона с наибольшим количеством работодателей, но без вакансий
query_5_2 = '''
select a.name, --Название региона
       count(e.area) emp_area, --Количество работодателей в регионе
       count(v.area_id) vac_area --регион вакансии
from employers e 
left join areas a on e.area = a.id --присоединяем таблицу регионов к работодателям
left join vacancies v on v.area_id = e.area --присоединяем вакансии по area_id
group by a.name, a.id --группируем по названию региона и ID региона
having count(v.area_id) =0 --Фильтрация по регионам без вакансий  
order by emp_area DESC --сортировка по убыванию количества работодателей 
'''

In [ ]:
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_720\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


name  emp_area  vac_area
0                          Россия       410         0
1                       Казахстан       207         0
2              Московская область        75         0
3              Краснодарский край        19         0
4                        Беларусь        18         0
..                            ...       ...       ...
181                       Светлый         1         0
182              Звёздный Городок         1         0
183                    Иноземцево         1         0
184  Котовск (Тамбовская область)         1         0
185             Псковская область         1         0

[186 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [6]:
# Запрос для определения количества регионов, в которых каждый работодатель публикует вакансии
query_5_3 = '''
select e.id, --ID работодателя
       count(distinct v.area_id) as cnt_area_id --количество уникальных id регионов
from employers e 
join vacancies v on v.employer_id = e.id --присоединение ID работодателей из таблицы vacancions = ID работодателей из таблицы employers
group by e.id --группировка по ID работодателей
order by count(distinct v.area_id) desc --сортировка по кол-ву уникальных id регионов
'''

In [7]:
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_12092\2398953733.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


id  cnt_area_id
0         1740          181
1         2748          152
2      5724811          116
3      5130287           88
4      3682876           71
...        ...          ...
14901   810278            1
14902   810313            1
14903   810551            1
14904   810688            1
14905   811902            1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [ ]:
# Запрос для поиска работодателей, не связанных ни с одной отраслью
query_5_4='''
select e.id, count(distinct ei.industry_id)
from employers e 
left join employers_industries ei on e.id = ei.employer_id --присоединение таблицы работодателей к доп. таблице сфер деятельности работодателей
group by e.id --группировка по ID-работодателей
having count(distinct ei.industry_id) = 0 --фильтрация по количеству 0
'''

In [ ]:
display(f'Количество работодателей без сфер деятельности: {pd.read_sql_query(query_5_4,connection).shape[0]}')

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_12092\1982987551.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(f'Количество работодателей без сфер деятельности:{pd.read_sql_query(query_5_4,connection).shape[0]}')


'Количество работодателей без сфер деятельности:8419'

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [13]:
#Запрос для поиска работодателей, находящемся на 3 месте, связанных ровно с 4 отраслями деятельности
query_5_5 ='''
select e.name, --название работодателя 
       count(ei.industry_id) --кол-во сфер деятельности
from employers e
left join employers_industries ei on e.id = ei.employer_id --присоединение таблицы по id-работодателя 
group by e.name --группировка по названию работодателя
having count(ei.industry_id)=4 --фильтрация по количеству сфер деятельности
order by e.name --группировка по названию
'''

In [21]:
display(f'Название компании, которая находиться на 3м месте(по названию):{pd.read_sql_query(query_5_5, connection).iloc[2,0]}')

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_12092\447132296.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(f'Название компании, которая находиться на 3м месте(по названию):{pd.read_sql_query(query_5_5, connection).iloc[2,0]}')


'Название компании, которая находиться на 3м месте(по названию):2ГИС'

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [24]:
# Запрос для подсчета количества работодателей в конкретной сфере деятельности - "Разработка программного обеспечения"
query_5_6= '''
select count(e.id), --количество работодателей  
       i.name --название сферы деятельности
from employers e 
join employers_industries ei on e.id = ei.employer_id  --присоединение employers_industries по id-работодателя
join industries i on i.id = ei.industry_id --присоединение industries по id-сферы деятельности
where i.name = 'Разработка программного обеспечения' --фильтрация по названию конкретной сферы деятельности
group by i.name --группировка по названию сферы деятельности
'''

In [27]:
display(f'Количество работодателей в сфере "Разработка программного обеспечения":{pd.read_sql_query(query_5_6,connection).iloc[0,0]}')

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_12092\1700392892.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(f'Количество работодателей в сфере "Разработка программного обеспечения":{pd.read_sql_query(query_5_6,connection).iloc[0,0]}')


'Количество работодателей в сфере "Разработка программного обеспечения":3553'

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [30]:
# Парсер wiki
# Импортируем библиотеку BeautifulSoup
from bs4 import BeautifulSoup
# Импортируем библиотеку requests
import requests
# url страница Википедии со списком городов миллионников России
url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'
# Выполняем GET запрос к указанному url и получаем ответ сервера 
response = requests.get(url)
# Создаем объект BeautifulSoup для парсинга HTML-контента
# response.text содержит HTML-код страницы, 'html.parser' - указание использовать встроенный парсер HTML
page = BeautifulSoup(response.text, 'html.parser')
# Создаем пустой список для хранения названий городов
city_list=list()
# Проходимся по всем строкам (tr) в теле таблицы (tbody)
# page.table - первая таблица на странице, .tbody - тело таблицы
for tr_row in page.table.tbody:
     # В каждой строке находим первую ссылку (тег <a>) и извлекаем ее текст (название города)
    city_list.append(tr_row.a.text)
# Удаляем первый элемент списка, так как он обычно содержит заголовок таблицы
# [1:] срез списка, который возвращает все элементы кроме первого (с индексом 0)
city_list = city_list[1:]

In [31]:
query_5_7 = f'''
(select a.name as city, count(*) as cnt 
from employers e
JOIN vacancies v on e.id = v.employer_id
JOIN areas a on a.id = v.area_id
WHERE e.name = 'Яндекс' and a.name in {tuple(city_list)}
group by a.id)
union all
(select 'Total', count(*) as cnt
FROM employers e
JOIN vacancies v on e.id = v.employer_id
JOIN areas a on a.id = v.area_id
WHERE e.name = 'Яндекс' and a.name in {tuple(city_list)})
'''

In [32]:
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_12092\3129374555.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


city  cnt
0             Total  485
1            Москва   54
2   Санкт-Петербург   42
3      Екатеринбург   39
4       Новосибирск   35
5         Волгоград   24
6           Воронеж   32
7         Краснодар   30
8        Красноярск   23
9   Нижний Новгород   36
10             Омск   21
11            Пермь   25
12   Ростов-на-Дону   25
13           Самара   26
14           Казань   25
15              Уфа   26
16        Челябинск   22

***

## Выводы по анализу работодателей:
- Яндекс находиться на первом месте по количеству вакансий, дальше идет Ростелеком, Тинькофф, СБЕР и Газпром нефть

- В регионе 'Россия' вакансий нет, но есть наибольшее количество работодателей. Логчно предположить, что работодатели регион не указывали, а указывали только страну. Либо же работа удаленная по всей России

- Максимальное число вакансий в различных регионах от одного работодателя - 181

- Количество работодателей без сфер деятельности 8419

- Работодатель находящийся на 3 месте(по названию), у которого указано 4 сферы деятельности - 2ГИС

- Количество работодателей в сфере "Разработка программного обеспечения": 3553

- Количество вакансий "Яндекса" в городах-миллионниках -485. Больше всего в Москве, чуть меньше в Санкт-Петербурге



# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [40]:
query_6_1  = f'''
select count(name) --количество вакансий
from vacancies v
where name ilike '%data%' or name ilike '%данн%' --фильтрация по содежранию в названии "data" & "данн"
'''


In [41]:
display(f'Количество вакансий, связанных с данными: {pd.read_sql_query(query_6_1,connection).iloc[0,0]}')

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_12092\3793004247.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(f'Количество вакансий, связанных с данными: {pd.read_sql_query(query_6_1,connection).iloc[0,0]}')


'Количество вакансий, связанных с данными: 1771'

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
# Запрос на подходящие вакансии для начинающего DS 
query_6_2 = f'''
select count(*) --количество строк 
from vacancies
where (name ilike '%data science%' or --фильтрация по содежранию в названии "data science" ИЛИ
       name ilike '%исследователь данных%' or --фильтрация по содежранию в названии  'исследователь данных' ИЛИ
       (name like '%ML%' and name not ilike '%HTML%') or --фильтрация по содежранию в названии 'ML', но не 'HTML' ИЛИ
       name ilike '%machine learning%' or  --фильтрация по содежранию в названии 'machine learning' ИЛИ
       name ilike '%машинн%обучен%' or --фильтрация по содежранию в названии 'машинн%обучен' ИЛИ
       name ilike '%data scientist%') --фильтрация по содежранию в названии 'data scientist' ИЛИ
       and --И
       
       (name ilike '%junior%' or --фильтрация по содежранию в названии 'junior' ИЛИ
        experience ilike '%Нет опыта%' or  --фильтрация по содежранию в названии 'Нет опыта' ИЛИ 
        employment ilike '%Стажировка%') --фильтрация по содежранию в названии 'Стажировка' ИЛИ
'''

In [ ]:
display(f'Количество подходящих названий для начинающего DS: {pd.read_sql_query(query_6_2, connection).iloc[0,0]}')

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_12092\44721352.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(f'Количество подходящих названий для наичнающего DS: {pd.read_sql_query(query_6_2, connection).iloc[0,0]}')


'Количество подходящих названий для наичнающего DS: 51'

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
# Запрос вакансий для DS, в качестве навыка SQL или POSTGRES
# Первая часть фильтрации одна и та же
query_6_3 = f'''
select count(name) --кол-во вакансий
from vacancies
where (name ilike '%data science%' or 
       name ilike '%исследователь данных%' or 
       (name ilike '%ML%' and name not ilike '%HTML%') or 
       name ilike '%machine learning%' or 
       name ilike '%машинн%обучен%' or 
       name ilike '%data scientist%')
       and
       
       (key_skills ilike '%SQL%' or --фильтрация key_skills по содежранию в названии 'SQL' ИЛИ
       key_skills ilike '%postgres%') --фильтрация key_skills по содежранию в названии 'postgres'

'''

In [52]:
display(f'Количество вакансий для DS, в которых качество ключевого навыка указан SQL или postgres: {pd.read_sql_query(query_6_3, connection).iloc[0,0]}')

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_12092\487365991.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(f'Количество вакансий для DS, в которых качество ключевого навыка указан SQL или postgres: {pd.read_sql_query(query_6_3, connection).iloc[0,0]}')


'Количество вакансий для DS, в которых качество ключевого навыка указан SQL или postgres: 229'

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
# Запрос на популярность Python в требованиях работодателя к DS
# Первая часть фильтрации одна и та же
query_6_4 = f'''
select count(name) 
from vacancies
where (name ilike '%data science%' or
       name ilike '%исследователь данных%' or 
       (name like '%ML%' and name not ilike '%HTML%') or 
       name ilike '%machine learning%' or 
       name ilike '%машинн%обучен%' or 
       name ilike '%data scientist%')
       and 
       
       key_skills ilike '%python%'  --фильтрация key_skills по содежранию в названии 'python'

'''

In [54]:
display(f'Количество вакансий в которых ключевые навыки содержит Python: {pd.read_sql_query(query_6_4, connection).iloc[0,0]}')


C:\Users\arinathebest\AppData\Local\Temp\ipykernel_12092\3297398764.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(f'Количество вакансий в которых ключевые навыки содержит Python: {pd.read_sql_query(query_6_4, connection).iloc[0,0]} ')


'Количество вакансий в которых ключевые навыки содержит Python: 351 '

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [56]:
# Запрос для расчета среднего количества ключевых навыков в вакансиях Data Science
# Вторая часть фильтрации = Первая часть фильтрации в предыдущих заданиях
query_6_5 = f'''
select round( --округление до 2х знаков после точки-разделителя
       avg( --среднее значение
       length(key_skills)- length(replace(key_skills, CHR(9), ''))+1 --вычитание общей длинны от количества разделителей + 1(кол-во навыков=кол-во разделителей + 1)
       ),
              2)
                    as avg_key_skills --псевдоним
from vacancies
where (name ilike '%data science%' or 
       name ilike '%исследователь данных%' or 
       (name like '%ML%' and name not ilike '%HTML%') or 
       name ilike '%machine learning%' or 
       name ilike '%машинн%обучен%' or 
       name ilike '%data scientist%')
'''

In [58]:
display (f'Количество ключевых навыков в вакансиях DS: {pd.read_sql_query(query_6_5, connection).iloc[0,0]}')


C:\Users\arinathebest\AppData\Local\Temp\ipykernel_12092\4113017109.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display (f'Количество ключевых навыков в вакансиях DS:{pd.read_sql_query(query_6_5, connection).iloc[0,0]}')


'Количество ключевых навыков в вакансиях DS:6.41'

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [63]:
query_6_6 = f'''
select experience, round( --округление до целого числа
       avg( --среднее значение
              coalesce( --функция возвращает первое не-NULL значение из списка
                     (salary_from + salary_to) / 2, --среднее арифметическое вилки зарплат 
                     salary_from, salary_to -- если нет salary_from -берем salary_to, если нет salary_to - берем salary_from
                     )
              ), 
                            0) 
                            as avg_salary --псевдоним  
from vacancies v
where (name ilike '%data science%' or
       name ilike '%исследователь данных%' or 
       (name like '%ML%' and name not like '%HTML%') or 
       name ilike '%machine learning%' or 
       name ilike '%машинн%обучен%' or 
       name ilike '%data scientist%')
       and 
       (salary_from is not null or salary_to is not null) --Если salary_from или salary_to не Null значение
group by experience --группировка по опыту работы
       '''

In [66]:
display(f'В среднем DS с опытом работы от 3 до 6 лет может расчитывать на зарплату: {pd.read_sql_query(query_6_6, connection).iloc[2,1]}')

C:\Users\arinathebest\AppData\Local\Temp\ipykernel_12092\3775905864.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  display(f'В среднем DS с опытом работы от 3 до 6 лет может расчитывать на зарплату: {pd.read_sql_query(query_6_6, connection).iloc[2,1]}')


'В среднем DS с опытом работы от 3 до 6 лет может расчитывать на зарплату: 243115.0'

***

# выводы по предметному анализу:

- Количество вакансий, связанных с данными: 1771

- Количество подходящих названий для начиающего DS: 51

- Количество вакансий для DS, в которых качество ключевого навыка указан SQL или postgres: 229.13% вакансий с ключевым навыком sql или postgres. Что делает этот навык очень важным

- Количество вакансий в которых ключевые навыки содержит Python-351.  20% вакансий включают в себя знания python. Что делает его must-have для специалистов

- Количество ключевых навыков в вакансиях DS: 6.41

- В среднем DS с опытом работы от 3 до 6 лет может расчитывать на зарплату: 243115.0

# Общий вывод по проекту

# Подведем итог исследования, обобщите выводы:

- Вакансий в базе 49тыс, а связанных с данными 1771. Соотвественно это нишевая область

- Количество вакансий для начинающего DS - 51. Что делает эту специальность не лучшим стартом в IT

- В вакансиях ЗП в промежутке от 71тыс до 110тыс, а требуемый опыт от 1 года до 3х лет. 

- Количество ключевых навыков - 6.41, среди них must-have являются python & sql/prosgres